simperby/src/lib.rs

```diff
@@ -86,9 +92,81 @@ pub struct ConsensusVoteItem {
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
 pub struct GenesisInfo {
     pub header: BlockHeader,
+    pub genesis_signature: FinalizationProof,
     pub chain_name: String,
 }
 
+impl GenesisInfo {
+    fn create_genesis_block(&self) -> Block {
+        Block {
+            header: self.header.clone(),
+            transactions: vec![],
+        }
+    }
+}
+
+/// A set of operations that may update the Simperby node state.
+///
+/// `ProposeBlock`, `SubmitConsensusVote` are explciitly performed through the `SimperbyApi` trait, and the others
+/// are done implicitly by the background task which is triggered by incoming p2p network messages.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub enum SimperbyOperation {
+    ProposeBlock {
+        block: Block,
+        signature: TypedSignature<BlockHeader>,
+    },
+    SubmitConsensusVote {
+        hash: Hash256,
+        signature: TypedSignature<BlockHeader>,
+    },
+    ReceiveProposal {
+        block: Block,
+        round: Round,
+        author_prevote: TypedSignature<(BlockHeader, Round)>,
+    },
+    ReceivePrevote {
+        block_hash: Hash256,
+        round: Round,
+        author_prevote: TypedSignature<(BlockHeader, Round)>,
+    },
+    ReceivePrecommit {
+        block_hash: Hash256,
+        round: Round,
+        author_prevote: TypedSignature<(BlockHeader, Round)>,
+    },
+    /// Used for sync.
+    ReceiveFinalizedBlock {
+        block: Block,
+        finalization_proof: Vec<(PublicKey, TypedSignature<BlockHeader>)>,
+    },
+}
+
+/// An error that may occur while accessing the Simperby node.
+#[derive(Error, Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub enum SimperbyError {
+    #[error("invalid block: {0}")]
+    InvalidBlock(String),
+    /// When the operation arguments are not valid.
+    #[error("invalid operation: {0}")]
+    InvalidOperation(String),
+    #[error("storage error: {0}")]
+    StorageError(simperby_kv_storage::Error),
+    /// When the storage is corrupted.
+    #[error("storage integrity error: {0}")]
+    StorageIntegrityError(String),
```

Comment:
What is the difference between the StorageError and StorageIntegrityError?

Reply:
StorageError: error from the raw storage
StorageIntegrityError: something was miss-stored (malformed data)

Reply:
How about adding a comment for `StorageError`?

Reply:
https://github.com/postech-dao/simperby/pull/51/files#diff-8b6508e21de37aeda2badaea3fc3756d498718ebe4dc8a5c698cefce8a8d761eR10
Also how about maintaining the same comments for StorageIntegrityError?

Reply:
fixed